hie-plugin-api/Haskell/Ide/Engine/Cradle.hs

```diff
@@ -534,25 +538,78 @@ cabalHelperCradle file = do
 -- The given FilePath must be relative to the Root of the project
 -- the given units belong to.
 getComponent
-  :: QueryEnv pt -> [Unit pt] -> FilePath -> IO (Maybe ChComponentInfo)
-getComponent _env [] _fp = return Nothing
-getComponent env (unit : units) fp =
-  try (runQuery (unitInfo unit) env) >>= \case
-    Left (e :: IOException) -> do
-      warningm $ "Catching and swallowing an IOException: " ++ show e
-      warningm
-        $  "The Exception was thrown in the context of finding"
-        ++ " a component for \""
-        ++ fp
-        ++ "\" in the unit: "
-        ++ show unit
-      getComponent env units fp
-    Right ui -> do
-      let components = M.elems (uiComponents ui)
-      debugm $ "Unit Info: " ++ show ui
-      case find (fp `partOfComponent`) components of
-        Nothing -> getComponent env units fp
-        comp    -> return comp
+  :: forall pt. QueryEnv pt -> [Unit pt] -> FilePath -> IO (Either [String] ChComponentInfo)
+getComponent env unitCandidates fp = getComponent' [] [] unitCandidates >>=
+    \case
+      (tried, failed, Nothing) -> return (Left $ buildErrorMsg tried failed)
+      (_, _, Just comp) -> return (Right comp)
+  where
+    getComponent' :: [UnitInfo] -> [(Unit pt, IOException)] -> [Unit pt] -> IO ([UnitInfo], [(Unit pt, IOException)], Maybe ChComponentInfo)
+    getComponent' triedUnits failedUnits [] = return (triedUnits, failedUnits, Nothing)
+    getComponent' triedUnits failedUnits (unit : units) =
+      try (runQuery (unitInfo unit) env) >>= \case
+        Left (e :: IOException) -> do
+          warningm $ "Catching and swallowing an IOException: " ++ show e
+          warningm
+            $  "The Exception was thrown in the context of finding"
+            ++ " a component for \""
+            ++ fp
+            ++ "\" in the unit: "
+            ++ show unit
+          getComponent' triedUnits ((unit, e):failedUnits) units
+        Right ui -> do
+          let components = Map.elems (uiComponents ui)
+          debugm $ "Unit Info: " ++ show ui
+          case find (fp `partOfComponent`) components of
+            Nothing -> getComponent' (ui:triedUnits) failedUnits units
+            comp    -> return (triedUnits, failedUnits, comp)
+
+    buildErrorMsg :: [UnitInfo] -> [(Unit pt, IOException)] -> [String]
+    buildErrorMsg triedUnits failedUnits =
+            [ "Could not obtain flags for: \"" ++ fp ++ "\"."
+            , ""
+            ]
+            ++ concat
+              [
+                [ "This Module was not part of any component we are aware of."
+                , ""
+                ]
+                ++ concatMap ppShowUnitInfo triedUnits
+                ++ [ ""
+                   , ""
+                   , "If you dont know how to expose a module, take a look at:"
+                   , "https://www.haskell.org/cabal/users-guide/developing-packages.html"
```

Comment:
Afaiu this will be called if we are using stack as well so maybe we should replace this link with a build tool agnostic one (or add an stack one)

Reply:
Both use the underlying cabal file format, so this link should be fine right? Stack doesn't provide any of its own documentation for it does it?

Reply:
Well, it has its own documentation, covering `package.yaml` and `stack.yaml`. Thanks to some of the lastest issues i am discovering that there are quite users using that combo and not sure if all of them even cares about the `.cabal` file. Moreover all cli examples are with cabal and they are not applicable with stack.

I think the more similar stack doc link could be: https://docs.haskellstack.org/en/stable/GUIDE/